✨ Add bootCommands to cloud-init file generation

davidumea opened the pull request:
<!-- Thanks for sending a pull request! Here are some tips for you:
    1. If this is your first time, please read our contributor guidelines: https://github.com/kubernetes-sigs/cluster-api/blob/main/CONTRIBUTING.md#contributing-a-patch and developer guide https://github.com/kubernetes-sigs/cluster-api/blob/main/docs/book/src/developer/getting-started.md

    2. Please add an icon to the title of this PR (see https://sigs.k8s.io/cluster-api/CONTRIBUTING.md#contributing-a-patch), and delete this line and similar ones
    the icon will be either ⚠️ (:warning:, major or breaking changes), ✨ (:sparkles:, feature additions), 🐛 (:bug:, patch and bugfixes), 📖 (:book:, documentation or proposals), or 🌱 (:seedling:, minor or other) 
-->

**What this PR does / why we need it**:
Adds the ability to provide [bootcmd](https://cloudinit.readthedocs.io/en/latest/reference/examples.html#run-commands-on-first-boot) commands to cloud-init via the KubeadmConfig or KubeadmConfigTemplates custom resources. (Same for KubeadmControlPlane)

~~One thing that I know that **does not work** at the time of creating this PR is validating/mutating webhooks for KubeadmConfig and KubeadmConfigTemplates. Some pointers to where to change that would be appreciated.~~

**Which issue(s) this PR fixes**
Fixes #

<!-- 
Please label this pull request according to what area(s) you are addressing. For reference on PR/issue labels, see: https://github.com/kubernetes-sigs/cluster-api/labels?q=area+

Area example:
/area runtime-sdk
-->

chrischdi commented:
Are there version constraints around `bootcmd`? Does it already exist in cloud-init for a long time?

What happens if we render it but the cloud-init version in the machine does not support it? Does it make cloud-init fail as a whole or is it just ignored?

davidumea commented:
> Are there version constraints around `bootcmd`? Does it already exist in cloud-init for a long time?
> 
> What happens if we render it but the cloud-init version in the machine does not support it? Does it make cloud-init fail as a whole or is it just ignored?
> 
> Edit: looks like it is already there for a long time (from the start?) at `0.7.7` https://cloudinit.readthedocs.io/en/0.7.7/topics/examples.html

Exactly I don't think that should be an issue as `bootcmd` has been part of cloud-init for a long time.

fabriziopandini commented:
/ok-to-test

cc @enxebre @vincepri @randomvariable

randomvariable commented on bootstrap/kubeadm/api/v1beta1/kubeadmconfig_types.go:
```diff
@@ -75,6 +75,10 @@ type KubeadmConfigSpec struct {
 	// +optional
 	Mounts []MountPoints `json:"mounts,omitempty"`
 
+	// BootCommands specifies extra commands to run very early in the boot process
+	// +optional
```
```suggestion
	// BootCommands specifies extra commands to run very early in the boot process via the cloud-init bootcmd
	// module. This is typically run in the cloud-init.service systemd unit. This has no effect in Ignition.
	// +optional
```

Make this super-explicit, also indicates to Ignition users that it won't be used there.

davidumea replied:
Thank you for the suggestion, I agree that it could've been clearer :+1:

randomvariable commented on bootstrap/kubeadm/api/v1beta1/kubeadmconfig_types.go:
```diff
@@ -75,6 +75,10 @@ type KubeadmConfigSpec struct {
+	// +optional
+	BootCommands []BootCommand `json:"bootCommands,omitempty"`
+
 	// PreKubeadmCommands specifies extra commands to run before kubeadm runs
 	// +optional
```
```suggestion
	// PreKubeadmCommands specifies extra commands to run before kubeadm runs. 
	// With cloud-init, this is prepended to the runcmd module configuration, and is typically executed in
	// the cloud-final.service systemd unit. In Ignition, this is prepended to /etc/kubeadm.sh.
	// +optional
```

And repeat for PostKubeadmCommands, replacing "prepended" with "appended".

randomvariable reviewed (changes requested):
Bootcmd module has been in cloud-init for as long as I can remember (which is back to 2010), so don't see an issue there.

I would like the documentation to be more explicit though, particularly about how this will have no effect on Ignition. Have suggested some changes.

neolit123 commented on internal/apis/bootstrap/kubeadm/v1alpha3/kubeadmconfig_types.go:
```diff
@@ -59,11 +59,20 @@ type KubeadmConfigSpec struct {
+	// BootCommands specifies extra commands to run very early in the boot process via the cloud-init bootcmd
+	// module. This is typically run in the cloud-init.service systemd unit. This has no effect in Ignition.
+	// +optional
+	BootCommands []BootCommand `json:"bootCommands,omitempty"`
```
shouldn't it be added only to v1beta1 which is the actively maintained API version?

davidumea replied:
If I remove this and run `make verify`, it removes `bootCommands` from the CRDs.

Might be an issue in the verify logic, what do you think?

```diff
diff --git a/bootstrap/kubeadm/config/crd/bases/bootstrap.cluster.x-k8s.io_kubeadmconfigs.yaml b/bootstrap/kubeadm/config/crd/bases/bootstrap.cluster.x-k8s.io_kubeadmconfigs.yaml
index 79c3b19e0..4e1852f77 100644
--- a/bootstrap/kubeadm/config/crd/bases/bootstrap.cluster.x-k8s.io_kubeadmconfigs.yaml
+++ b/bootstrap/kubeadm/config/crd/bases/bootstrap.cluster.x-k8s.io_kubeadmconfigs.yaml
@@ -47,17 +47,6 @@ spec:
               KubeadmConfigSpec defines the desired state of KubeadmConfig.
               Either ClusterConfiguration and InitConfiguration should be defined or the JoinConfiguration should be defined.
             properties:
-              bootCommands:
-                description: |-
-                  BootCommands specifies extra commands to run very early in the boot process via the cloud-init bootcmd
-                  module. This is typically run in the cloud-init.service systemd unit. This has no effect in Ignition.
-                items:
-                  description: BootCommand defines input for each bootcmd command
-                    in cloud-init.
-                  items:
-                    type: string
-                  type: array
-                type: array
               clusterConfiguration:
                 description: ClusterConfiguration along with InitConfiguration are
                   the configurations necessary for the init command
```
etc

fabriziopandini replied:
New fields should not be added to older API versions.
Conversions should be implemented instead.
See eg https://github.com/kubernetes-sigs/cluster-api/blob/3db18fb824790c17a294b3a92dd147a174bf1dc9/internal/apis/bootstrap/kubeadm/v1alpha3/conversion.go#L47-L56

davidumea replied:
> If I remove this and run make verify, it removes bootCommands from the CRDs.
Disregard this, I didn't realize that all versions of the crds were in the same manifest.

fabriziopandini commented on bootstrap/kubeadm/internal/cloudinit/cloudinit_test.go:
```diff
@@ -34,6 +34,7 @@ func TestNewInitControlPlaneAdditionalFileEncodings(t *testing.T) {
 	cpinput := &ControlPlaneInput{
 		BaseUserData: BaseUserData{
 			Header:              "test",
+			BootCommands:        nil,
```
Can we have a test for BootCommands not nil for init, join CP, join workers?

davidumea replied:
I added non-nil values for `TestNewInitControlPlaneCommands` and `TestNewJoinControlPlaneAdditionalFileEncodings`, did you want me to add more tests or is this enough? I couldn't find one for "join workers" https://github.com/kubernetes-sigs/cluster-api/pull/11271/commits/daf4f97e56614c62bec80c80635f094a81cfca33

fabriziopandini commented on controlplane/kubeadm/internal/filters_test.go:
```diff
@@ -548,7 +548,7 @@ func TestMatchInitOrJoinConfiguration(t *testing.T) {
     JoinConfiguration: nil,
     Files:             nil,
-    ... // 10 identical fields
+    ... // 11 identical fields
```
I need to take some time to dig into impact of this change on existing clusters, because we need to make sure this change does not trigger rollouts
However, I'm not sure when I will get to it 😓

davidumea replied:
Do you think it's feasible to have this merged before the next minor release?

sbueringer replied:
Yup. All good.

fabriziopandini commented on bootstrap/kubeadm/api/v1beta1/kubeadmconfig_types.go:
```diff
@@ -75,11 +75,20 @@ type KubeadmConfigSpec struct {
 	Mounts []MountPoints `json:"mounts,omitempty"`
 
-	// preKubeadmCommands specifies extra commands to run before kubeadm runs
+	// BootCommands specifies extra commands to run very early in the boot process via the cloud-init bootcmd
```
FYI during this cycle we started optimizing go doc comments in API packages for usage with kubectl explain, so yaml/json names should be used
```suggestion
	// bootCommands specifies extra commands to run very early in the boot process via the cloud-init bootcmd
```

Same for preKubeadmCommands and postKubeadmCommands
(and same in all the others API packages touched by this PR)

fabriziopandini commented on bootstrap/kubeadm/internal/cloudinit/cloudinit_test.go:
```diff
@@ -94,6 +95,9 @@ func TestNewInitControlPlaneCommands(t *testing.T) {
 	cpinput := &ControlPlaneInput{
 		BaseUserData: BaseUserData{
+			BootCommands: []bootstrapv1.BootCommand{
+				{"echo", "hello"},
```
In order to make this test to be useful to validate the code actually works/to prevent regressions, you should also modify test assertions down below that boot commands actually show up in the generated output (and possibly, ensuring that command show up in the right section of the generated file)

davidumea replied:
I wasn't sure how to ensure that the commands show up in the right section of the generated file, but I added a check that the bootcommands show up in the generated file and also that the `bootcmd:` string is there (which is omitted when bootCommands are unset)

https://github.com/kubernetes-sigs/cluster-api/pull/11271/commits/a4f3bdbc53b27837ca95cd7ca67419b2e0afe85d

What do you think?

davidumea replied:
Thanks for the suggestion!

https://github.com/kubernetes-sigs/cluster-api/pull/11271/commits/6fb468d53a18804bc2b585138a12a68f008efa7e

chrischdi commented:
/test pull-cluster-api-e2e-main

chrischdi commented:
/test pull-cluster-api-e2e-main

chrischdi commented:
/retest

Looks like a flake in creating the docker container and overlapping ports

chrischdi commented:
/retest

Just to make sure its really not a flake

davidumea commented:
@fabriziopandini Can you take another look? I added some new tests and addressed the comments

Danil-Grigorev commented on bootstrap/kubeadm/internal/cloudinit/boot_commands.go:
```diff
@@ -0,0 +1,29 @@
+/*
+Copyright 2019 The Kubernetes Authors.
```
```suggestion
Copyright 2025 The Kubernetes Authors.
```
Probably a nitpick, but it looks a bit outdated.

Danil-Grigorev commented on controlplane/kubeadm/internal/webhooks/kubeadm_control_plane.go:
```diff
@@ -208,6 +209,7 @@ func (webhook *KubeadmControlPlane) ValidateUpdate(_ context.Context, oldObj, ne
 		{spec, kubeadmConfigSpec, joinConfiguration, "discovery"},
 		{spec, kubeadmConfigSpec, joinConfiguration, "discovery", "*"},
 		// spec.kubeadmConfigSpec
+		{spec, kubeadmConfigSpec, "bootCommands"},
```
```suggestion
		{spec, kubeadmConfigSpec, bootCommands},
```

davidumea replied:
Thanks for finding this!

Danil-Grigorev commented on bootstrap/kubeadm/api/v1beta1/kubeadmconfig_types.go:
```diff
@@ -75,11 +75,20 @@ type KubeadmConfigSpec struct {
 
-	// preKubeadmCommands specifies extra commands to run before kubeadm runs
+	// bootCommands specifies extra commands to run very early in the boot process via the cloud-init bootcmd
+	// module. This is typically run in the cloud-init.service systemd unit. This has no effect in Ignition.
```
If there is no equivalent for the ignition format, it is good to prevent user confusion by introducing a validation webhook check on `kubeadmConfig.spec.format == “cloud-init”` for this field to be set.

Danil-Grigorev replied:
`bootstrap/kubeadm/internal/webhooks/kubeadmconfig.go` seems to be the place for validation webhooks code

davidumea replied:
Do you mean to validate that `format == "clout-init"` only if ignition is not used? Is that based on `feature.KubeadmBootstrapFormatIgnition` or is there another way of checking if ignition should be used?

Danil-Grigorev replied:
Format field on the config designates the type of configuration used. If it is cloud-init, then it is not ignition :)

davidumea replied:
Okay thanks, is this how you meant? https://github.com/kubernetes-sigs/cluster-api/pull/11271/commits/b7c575a63f5f0aafbb6108076859ac36db7d6d8e

Also, I believe the [default value for the format](https://github.com/kubernetes-sigs/cluster-api/blob/6fb468d53a18804bc2b585138a12a68f008efa7e/bootstrap/kubeadm/api/v1beta1/kubeadmconfig_types.go#L35) is "cloud-config", but please let me know if I misunderstood and it should be "cloud-init".

anmazzotti commented on bootstrap/kubeadm/internal/webhooks/kubeadmconfig.go:
```diff
@@ -65,6 +65,9 @@ func (webhook *KubeadmConfig) ValidateCreate(_ context.Context, obj runtime.Obje
 	if !ok {
 		return nil, apierrors.NewBadRequest(fmt.Sprintf("expected a KubeadmConfig but got a %T", obj))
 	}
+	if c.Spec.BootCommands != nil && c.Spec.Format != "cloud-config" {
```
May you use the [constant](https://github.com/kubernetes-sigs/cluster-api/blob/main/bootstrap/kubeadm/api/v1beta1/kubeadmconfig_types.go#L35) instead?

davidumea replied:
https://github.com/kubernetes-sigs/cluster-api/pull/11271/commits/0484cdc93ec924e4c5699a05921a89fdd7a5e4b8

davidumea replied:
Thank you! I have now crossed out the issue I mentioned in the initial PR description, since the webhooks now works as expected in my testing.

> One thing that I know that does not work at the time of creating this PR is validating/mutating webhooks for KubeadmConfig and KubeadmConfigTemplates. Some pointers to where to change that would be appreciated.

anmazzotti commented on bootstrap/kubeadm/internal/webhooks/kubeadmconfig.go:
```diff
@@ -75,6 +78,9 @@ func (webhook *KubeadmConfig) ValidateUpdate(_ context.Context, _, newObj runtim
 	if !ok {
 		return nil, apierrors.NewBadRequest(fmt.Sprintf("expected a KubeadmConfig but got a %T", newObj))
 	}
+	if newC.Spec.BootCommands != nil && newC.Spec.Format != "cloud-config" {
```
Same here for the constant usage

anmazzotti commented on bootstrap/kubeadm/internal/webhooks/kubeadmconfig_test.go:
```diff
@@ -457,6 +457,36 @@ func TestKubeadmConfigValidate(t *testing.T) {
+					Namespace: metav1.NamespaceDefault,
+				},
+				Spec: bootstrapv1.KubeadmConfigSpec{
+					Format: "cloud-config",
```
And here

anmazzotti reviewed (changes requested):
All looks good to me, beside the minor issue of not using the constant "cloud-config"

Danil-Grigorev commented on bootstrap/kubeadm/internal/webhooks/kubeadmconfig.go:
```diff
@@ -65,6 +65,9 @@ func (webhook *KubeadmConfig) ValidateCreate(_ context.Context, obj runtime.Obje
 	if !ok {
 		return nil, apierrors.NewBadRequest(fmt.Sprintf("expected a KubeadmConfig but got a %T", obj))
 	}
+	if c.Spec.BootCommands != nil && c.Spec.Format != bootstrapv1.CloudConfig {
```
It seems this commands need to be a part of `webhook.validate`, otherwise the check looks fine.

davidumea replied:
I had to update the `validateIgnition` function so that `allErrs` wouldn't be empty, which was probably for the best, thanks

https://github.com/kubernetes-sigs/cluster-api/pull/11271/commits/3db48fac929b60f078be4204d3df03211ff2e0bc

davidumea replied:
@Danil-Grigorev Please take a look if you have time :slightly_smiling_face:

Danil-Grigorev commented:
/lgtm

Seems good to me now, thank you

davidumea commented:
/assign justinsb

Pinging @justinsb for approval

chrischdi commented on bootstrap/kubeadm/internal/ignition/clc/clc_test.go:
```diff
@@ -73,6 +72,11 @@ func TestRender(t *testing.T) {
 		{
 			desc: "renders valid Ignition JSON",
 			input: &cloudinit.BaseUserData{
+				BootCommands: []bootstrapv1.BootCommand{
+					{
+						"boot-command", "another-boot-command",
+					},
+				},
```
This does not make sense to get set right? Because ignition does not have `bootCommands` and we block this getting set?

davidumea replied:
You're right, it doesn't make sense to have that there.

chrischdi commented on bootstrap/kubeadm/internal/cloudinit/cloudinit_test.go:
```diff
@@ -94,9 +95,12 @@ func TestNewInitControlPlaneCommands(t *testing.T) {
 	cpinput := &ControlPlaneInput{
 		BaseUserData: BaseUserData{
+			BootCommands: []bootstrapv1.BootCommand{
+				{"echo", "$(date) hello BootCommands!"},
```
Question: can't this be the same as below?

```suggestion
				{`"echo $(date) ': hello BootCommands!'"`},
```

davidumea replied:
I wanted to test the output when having multiple commands, but it didn't make sense to split it there.

This should be clearer https://github.com/kubernetes-sigs/cluster-api/pull/11271/commits/1874ba875fa80ef4afb1fe11786cb69d1829136d

chrischdi reviewed (commented):
/lgtm

davidumea commented:
@chrischdi Is there anything I can do for this to progress?

chrischdi commented on bootstrap/kubeadm/api/v1beta1/kubeadmconfig_types.go:
```diff
@@ -75,11 +75,20 @@ type KubeadmConfigSpec struct {
 	Mounts []MountPoints `json:"mounts,omitempty"`
 
-	// preKubeadmCommands specifies extra commands to run before kubeadm runs
+	// bootCommands specifies extra commands to run very early in the boot process via the cloud-init bootcmd
+	// module. This is typically run in the cloud-init.service systemd unit. This has no effect in Ignition.
```
I just learned that bootCommands run on every boot (`always`), I think that's worth to mention.

davidumea replied:
Sure, I wrote what they say in the [docs](https://cloudinit.readthedocs.io/en/latest/reference/examples.html#run-commands-on-first-boot)

chrischdi commented:
> @chrischdi Is there anything I can do for this to progress?

Thanks for asking, I'm sure this is in the backlog of some maintainers, at least this should also bring it up again in notifications if it got lost ;-)

fabriziopandini reviewed (commented):
Last question from my side, otherwise sgtm

davidumea commented on bootstrap/kubeadm/api/v1beta1/kubeadmconfig_types.go:
```diff
@@ -75,11 +75,21 @@ type KubeadmConfigSpec struct {
+	// module. bootcmd will run on every boot, 'cloud-init-per' command can be used to make bootcmd run exactly
+	// once. This is typically run in the cloud-init.service systemd unit. This has no effect in Ignition.
+	// +optional
+	BootCommands []BootCommand `json:"bootCommands,omitempty"`
```
At least my use case for this is writing `cloud-init-per` bootcommands, and in the [cloudinit documentation](https://cloudinit.readthedocs.io/en/latest/reference/examples.html#run-commands-on-first-boot) this is how they write those commands. Seems like both could be allowed, but I'm not sure it's worth spending time on that?

sbueringer replied:
Hm, but wouldn't []string also allow to render `[ cloud-init-per, once, mymkfs, mkfs, /dev/vdb ]`

I would really like to keep this consistent with pre / post kubeadm commands (which are rendered into run)

fabriziopandini replied:
> the [cloudinit documentation](https://cloudinit.readthedocs.io/en/latest/reference/examples.html#run-commands-on-first-boot) this is how they write those commands

I would not use python as a reference to how do we implement types in go

> I would really like to keep this consistent with pre / post kubeadm commands (which are rendered into run)

Same

davidumea replied:
It makes sense to use the same format as the other kubeadm commands.

https://github.com/kubernetes-sigs/cluster-api/pull/11271/commits/005b5075f6d571fb1b20c204b2a49f7f57297a8a

sbueringer replied:
Thank you!

sbueringer commented:
Just a heads up. Code freeze for 1.10 is on 18th March. So we maybe have 1,5 weeks left if this should make it into v1.10.

davidumea commented:
Seems like there has been some updates to the linter on main, will rebase and fix

sbueringer commented:
/test pull-cluster-api-e2e-main

sbueringer commented:
/assign

sbueringer commented:
@davidumea One very last comment. Feel free to squash and ping me afterwards and I'll merge

davidumea commented:
> @davidumea One very last comment. Feel free to squash and ping me afterwards and I'll merge

Sure, I added MinLength/MaxLength for bootcommands, rebased on main again and squashed.

When running `make test` locally the [TestNewClusterClient](https://github.com/kubernetes-sigs/cluster-api/blob/afd68cda45eb120b4fc9c62ab44ca719f0861682/controllers/remote/cluster_test.go#L87) test failed for me but it also failed when running it on main, so I figured it was unrelated to the changes in this PR.

sbueringer commented:
Agree, that seems unrelated.

Thank you very much for working on this. Sorry it took a while

/lgtm
/approve
